### ISSUE TITLE
aws-appsync-codegen generate failed

### ISSUE THREAD

**Issue by Derek-X-Wang** [awslabs/aws-appsync-codegen#3]
Hi,

I downloaded schema.graphql and schema.json directly from AppSync Sample schema.

When I run,

`aws-appsync-codegen generate schema.graphql --schema schema.json --output API.swift`

it failed with:

```
.../Users/derekxinzhewang/Downloads/schema.graphql: Directive "aws_subscribe" may not be used on FIELD_DEFINITION.
error: Validation of GraphQL query document failed
```

if I remove aws_subscribe directive, aws-appsync-codegen generation success, but only comment is generated. #2 same issue

thanks

**Comment by undefobj** [comment]
Hi @Derek-X-Wang we're currently looking into this.

**Comment by teddarling** [comment]
I can confirm seeing the same when trying to generate with subscribe types using different targets as well

**Comment by undefobj** [comment]
Hi @Derek-X-Wang it looks like you're using the SDL file which defines your GraphQL API and not an *.graphql file which has queries and mutations that define the selection set for your operations.

We have a detailed walk through of this process in the AWS AppSync documentation: https://docs.aws.amazon.com/appsync/latest/devguide/building-a-client-app-ios.html

Can you try again with GraphQL queries and/or mutations defined?

**Comment by teddarling** [comment]
I have issue #2 working (if I remove `@aws_subscribe` from the generated .graphql file the AppSync dashboard provides

Having said that. If I use schema.graphql file downloaded from the AppSync dashboard to start (needed for getting my types in place), then that file includes ` @aws_subscribe(mutations: ["createTemplateType"])` at the end of each line inside the Subscription type:

```
type Subscription {
  onCreateTemplateType(description: String, id: ID, imageKey: String, imageKeyHover: String, title: String): TemplateType @aws_subscribe(mutations: ["createTemplateType"])
}
```

Now, even though I can get the generation to work, using the file that AppSync provides as the starting point still causes this error to occur.

Below is an exact download of a schema.graphql file that AppSync created for me (It so happens to match what is in the schema editor):

```type Mutation {
  createTemplateType(input: CreateTemplateTypeInput!): TemplateType
  deleteTemplateType(input: DeleteTemplateTypeInput!): TemplateType
  updateTemplateType(input: UpdateTemplateTypeInput!): TemplateType
}

type Query {
  getItems: ID
  getTemplateType(id: ID!): TemplateType
  listTemplateTypes(after: String, first: Int): TemplateTypeConnection
}

type Subscription {
  onCreateTemplateType(description: String, id: ID, imageKey: String, imageKeyHover: String, title: String): TemplateType
    @aws_subscribe(mutations:["createTemplateType"])
  onDeleteTemplateType(description: String, id: ID, imageKey: String, imageKeyHover: String, title: String): TemplateType
    @aws_subscribe(mutations:["deleteTemplateType"])
  onUpdateTemplateType(description: String, id: ID, imageKey: String, imageKeyHover: String, title: String): TemplateType
    @aws_subscribe(mutations:["updateTemplateType"])
}

type TemplateType {
  description: String
  id: ID!
  imageKey: String
  imageKeyHover: String
  title: String!
}

type TemplateTypeConnection {
  items: [TemplateType]
  nextToken: String
}

input CreateTemplateTypeInput {
  description: String
  id: ID!
  imageKey: String
  imageKeyHover: String
  title: String!
}

input DeleteTemplateTypeInput {
  id: ID!
}

input UpdateTemplateTypeInput {
  description: String
  id: ID!
  imageKey: String
  imageKeyHover: String
  title: String
}
```

I guess that I would expect that the file provided by the AppSync dashboard, would at least generate (even if comments because no queries or mutations have been created) with the .graphql file that is offered for download.

**Comment by undefobj** [comment]
@teddarling the file offered in the dashboard is an exportable version of the schema to share between APIs. Many customers use these for their CI/CD flows. GraphQL by design is made so that the client developer defines their queries and mutations, along with the fields they want returned from this. We can look at some autogeneration of queries or mutations in the future, but even with that it's impossible to know what data you want returned to the client or the depth of your queries. 

The normal workflow is to define what your queries or mutations are in the "Queries" pane of the console. Then save these to a queries.graphql or a mutations.graphql file for codegen.

So to recap:
1. The schema.graphql file in the console is just the Schema definition file written in Schema Definition Language (SDL). This is NOT for codegen and only to serve as an API definition for future APIs or updates.
2. The schema.json file is the "introspection schema" and is needed for codegen as it defines how the client knows the shape of the API that was created
3. You define your queries and mutations as a client developer, which define the data that will be sent or retrieved over the network. You need this for codegen along with the schema.json file.

**Comment by teddarling** [comment]
Downloading the schema.json file works great. 

The issue then appears to be that it is best to avoid downloading schema.graphql from the AppSync console and running `introspect-schema` on it. Because the schema.json file created from that is where the error about `@aws_subscribe` is coming from.

**Comment by undefobj** [comment]
Correct - for clarity you should NOT run any commands on the schema.graphql file from the console, that's simply the SDL file for creating new APIs or updating them either using the console or AWS CLI. For codegen you should only download the introspection schema json file and provide a file(s) with your queries and mutations.

**Comment by teddarling** [comment]
Good info for me and I'm moving along fine now. I can't quite speak for @Derek-X-Wang 

Out of curiosity is there a place where we can send feature requests for what goes on within the console or for the new AppSync service in general?

**Comment by undefobj** [comment]
You can always use the "Feedback" button in the AWS AppSync console, or alternatively post to the AWS forums: https://forums.aws.amazon.com/forum.jspa?forumID=280&start=0

**Comment by Derek-X-Wang** [comment]
Great thanks for @undefobj @teddarling !

I misunderstood the different between SDL and .graphql files.

Codegen and aws_subscribe work now!

I'm new to GraphQL and following the instructions on aws-mobile-appsync-events-starter-ios. It was really natural to use the schema.graphql from the AppSync dashboard without thinking.

Thanks
